### PR TITLE
do not report use-before-declaration error in async like IIFE initial…

### LINKED ISSUE #30907: IIFEs should not be CFA-inlined if the FE is a generator function
This is an incorrect error because generators don't run on invocation

```ts
function wrapI() {
    const iter = (function* foo() {
        iter; // use before def error
        yield 1;
    })();
}
```

cc @bterlson

### PR DESCRIPTION
**Pull request by Zzzen**
…izer

Fixes #30907

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -3841,7 +3840,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
      */
     function isSameScopeDescendentOf(initial: Node, parent: Node | undefined, stopAt: Node): boolean {
         return !!parent && !!findAncestor(initial, n => n === parent
-            || (n === stopAt || isFunctionLike(n) && (!getImmediatelyInvokedFunctionExpression(n) || isAsyncFunction(n)) ? "quit" : false));
+            || (n === stopAt || isFunctionLike(n) && (!getImmediatelyInvokedFunctionExpression(n) || (getFunctionFlags(n) & (FunctionFlags.AsyncGenerator))) ? "quit" : false));
```
[the remark this answers is not in the document]
Comment (Zzzen) [reply src/compiler/checker.ts:3843]: It's strange that i'm getting error diff updated when nothing has changed. image

**Review by gabritto — APPROVED** [review]

**Comment by gabritto** [comment]
@Zzzen could you fix the merge conflicts? I couldn't push to your branch myself

**Comment by Zzzen** [comment]
Of course, fixed.

**Comment by gabritto** [comment]
@typescript-bot run DT
@typescript-bot user test this
@typescript-bot test top100
@typescript-bot perf test this

**Comment by gabritto** [comment]
@typescript-bot run DT
@typescript-bot user test this
@typescript-bot test top100
@typescript-bot perf test this

**Comment by gabritto** [comment]
@typescript-bot run DT

**Comment by gabritto** [comment]
@typescript-bot run DT

**Pull request merged**